tests/testsuite/deduplicate_workspace.rs

```diff
@@ -0,0 +1,861 @@
+            license = "MIT"
+            license-file = "LICENSE"
+            keywords = ["cli"]
+            categories = ["development-tools"]
+            publish = false
+            edition = "2018"
+
+            [workspace.badges]
+            gitlab = { repository = "https://gitlab.com/rust-lang/rust", branch = "master" }
+
+            [workspace.dependencies]
+            dep = "0.1"
+        "#,
+        )
+        .file(
+            "bar/Cargo.toml",
+            r#"
+              [package]
+              name = "bar"
+              version = "0.1.0"
+              authors = []
+              workspace = ".."
+              "#,
+        )
+        .file("bar/src/main.rs", "fn main() {}")
+        .build();
+
+    p.cargo("build")
+        // Should not warn about unused fields.
+        .with_stderr(
+            "\
+[COMPILING] bar v0.1.0 ([CWD]/bar)
+[FINISHED] dev [unoptimized + debuginfo] target(s) in [..]
+",
+        )
+        .run();
+
+    p.cargo("check").run();
+    let lockfile = p.read_lockfile();
+    assert!(!lockfile.contains("dep"));
+}
+
+#[cargo_test]
+fn deny_optional_dependencies() {
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+            cargo-features = ["workspace-inheritance"]
+
+            [workspace]
+            members = ["bar"]
+
+            [workspace.dependencies]
+            dep1 = { version = "0.1", optional = true }
+        "#,
+        )
+        .file("src/main.rs", "fn main() {}")
+        .file(
+            "bar/Cargo.toml",
+            r#"
+              [package]
+              name = "bar"
+              version = "0.1.0"
+              authors = []
+              workspace = ".."
+              "#,
+        )
+        .file("bar/src/main.rs", "fn main() {}")
+        .build();
+
+    p.cargo("build")
+        .with_status(101)
+        .with_stderr(
+            "\
+[ERROR] failed to parse manifest at `[..]foo/Cargo.toml`
+
+Caused by:
+  dep1 is optional, but workspace dependencies cannot be optional
+",
+        )
+        .masquerade_as_nightly_cargo()
+        .run();
+}
+
+#[cargo_test]
+fn inherit_own_workspace_fields() {
+    registry::init();
+
+    let p = project().build();
+
+    let _ = git::repo(&paths::root().join("foo"))
+        .file(
+            "Cargo.toml",
+            r#"
+            cargo-features = ["workspace-inheritance"]
+            badges = { workspace = true }
+
+            [package]
+            name = "foo"
+            version = { workspace = true }
+            authors = { workspace = true }
+            description = { workspace = true }
+            documentation = { workspace = true }
+            homepage = { workspace = true }
+            repository = { workspace = true }
+            license = { workspace = true }
+            keywords = { workspace = true }
+            categories = { workspace = true }
+            publish = { workspace = true }
+            edition = { workspace = true }
```

Comment:
Nit: with TOML now supporting dotted keys, should we stylistically be preferring `version.workspace = true`?

Doesn't need to be addressed now but what tone we set in rolling out this feature will likely be copied by people going forward.

Reply:
I personally think `version.workspace = true` looks better and is more descriptive on fields. I can go either way on this and will do whatever you recommend. I would have to change all instances of `{ workspace = true }` to `version.workspace = true` in comments and tests but this isn't a problem. I think we should leave it as `{ workspace = true }` for dependencies since they can also have `features` and `optional`.

I can make this change right now as I still have to push a commit for other fixes.

Reply:
I can go either way on dependencies.  The same could be said for `dep = "version"` but I also dislike having to change the structure of data to insert a field (similarly, I personally dislike `use` statements and especially `use {};`, thats just asking for merge conflicts)

Reply:
I think this is something we should add to a cleanup/optimization PR. Changing all `field = { workspace = true }` to `field.workspace = true` might be outside of the scope of this PR and is something that we should talk about in the RFC.

Reply:
I've put it in the tracking issue regarding documentation

Marking this as resolved.